README.md
# labtex
[![](https://img.shields.io/badge/Documentation-007ec6.svg)](https://www.cianlm.dev/labtex)
[![PyPI](https://img.shields.io/pypi/v/labtex)](https://pypi.org/project/labtex/)
![Tests](https://github.com/CianLM/labtex/actions/workflows/python-publish.yml/badge.svg)
![License](https://img.shields.io/github/license/cianlm/labtex)

New extensive [Documentation](https://www.cianlm.dev/labtex)!
## Purpose

This package provides a single solution to repetitive analysis tasks in a lab environment. If you are doing error propagation, linear regression or LaTeX tables/figures manually, this package automates the process.

## Features

- Measurement and MeasurementList classes with automatic:
    - Error propagation
    - Printing to correct significant figures
    - Unit parsing, propagation and conversion
- Linear regression and nonlinear curve fitting
- A Document class with automatic:
    - LaTeX table generation in two different styles
    - Plot generation with matplotlib
    - Insertion/updating of tables and figures into a LaTeX file

## Installation

Using `pip` or `pip3` the latest release can be installed with
```
pip3 install labtex
```

## Usage

For ease of use, you can import the package into your file's global namespace with
```python
from labtex import *
```
The rest of this section will assume the package is imported in this way. Alternatively do `import labtex as lt`.

Single measurements can be instantiated with `Measurement(value,uncertainty,unit)` where unit is a string that will be parsed.
```python
x = Measurement(1.1,0.3,"m")
y = Measurement(2.22,0.4,"m")
z = M(314,10,"V")
```
where `M` is an equivalent shorthand for `Measurement`. Note that the unit parsing supports all combinations of common units, prefixes and powers of units, eg. any of "nm^2", "C^-1", "kg m^2 s^-2", "J^3" etc. are supported.

Measurement instances support all operations (`+-*/` and `**`) as well as math functions with the error and units automatically propagated. Some examples are shown below.
```python
print(x)
# 1.1 ± 0.3 m

print(x + y)
# 3.3 ± 0.5 m

print(x * z)
# (35 ± 9) × 10^{1} V m

print(x ** 2)
# 1.2 ± 0.7 m^2

print(Measurement.tan(x / y))
# 0.5 ± 0.2  
```
Notice also that Measurements are rounded to the significant figures as dictated by the uncertainty.

For a list of measurements, the `MeasurementList` class functions identically to the `Measurement` class, only now taking a list of values. The uncertainty can be a list or a single value for all measurements.

```python
heights = MeasurementList([185,183,182,194,184,177],5,"cm")
# Note: One could use the shorthand class ML(values,uncertainty,unit)

print(heights)
# [185 ± 5, 183 ± 5, 182 ± 5, 194 ± 5, 184 ± 5, 177 ± 5] cm

print(200 - heights)
# [15 ± 5, 17 ± 5, 18 ± 5, 6 ± 5, 16 ± 5, 23 ± 5] cm
```
`MeasurementList`s also support all operations (`+-*/` and `**`) with themselves and with `Measurement`s. 

With two `MeasurementList` instances, they can be linearly regressed with the `LinearRegression` class.
```python
voltages = MeasurementList([1.3,3,5,7,8.5,10],1,"V")
temperatures = MeasurementList([23,55,67,82,88,96],[5,3,7,10,5,6],"K")

reg = LinearRegression(voltages,temperatures)

print(lobf)
# m = 7 ± 1 V^{-1} C
# c = 27 ± 7 C
```
Observe that printing all `Measurement` and `MeasurementList` instances rounds the value to the largest significant figure of the error, as is convention.

To visualise this linear regression, the `plot` method can be used.
```python
fig = linreg.plot(xlabel="Voltage, V", ylabel="Temperature, T", title="Voltage and Temperature Plot")
fig.show()
```
This generates the following plot, in which the line of best fit is shown, as well as the region of uncertainty (in the linear regression parameters). Notice that the units are automatically included in the axis labels.

![](https://github.com/CianLM/labtex/raw/master/figures/graph1.png)

In order to include such a figure or a table of the data used in a LaTeX document, the `Document` class can be used. Argument names are not required, they are shown here only for demonstration.
```python
doc = Document(title="Lab Report Template",author="CianLM", filename="report.tex")
```
Note that `filename` here can be a preexisting `.tex` file or a new one to be created by labtex.

Once `doc` has been instantiated, tables and figures may be added to the document with their respective methods. Once again argument names are for demonstration and most are optional anyway.

```python
doc.add_table(
    nameandsymbol = ["Voltage, V","Temperature, T"],
    data = [voltages,temperatures],
    caption = "Voltage and Temperature Correlation"
)
```
This inserts the following into the `doc` instance.
```latex
\begin{table}[ht]
    \centering
    \caption{Voltage and Temperature Correlation}
    \label{tab:1}
    \begin{tabular}{c|cccccc}
        \toprule
            Voltage, V, ($\pm 1$ V) & 1 & 3 & 5 & 7 & 8 & 10 \\ 
            Temperature, T& $23 \pm 5 $ & $55 \pm 3 $ & $67 \pm 7 $ & $(8 \pm 1) \times 10^{1} $ & $88 \pm 5 $ & $96 \pm 6 $ \\ 
        \bottomrule
    \end{tabular}
\end{table}
```
which results in

![](https://github.com/CianLM/labtex/raw/master/figures/readmetable.png)


Alternatively if an `upright` table is preferred, this may be specified through the `style` argument.
```python
doc.add_table(
    nameandsymbol = ["Voltage, V", "Temperature, T"],
    data = [voltages,temperatures**2],
    caption = "Voltage and Temperature Squared Correlation",
    style = "upright"
)
```

Once again, this inserts the following into the `doc` instance.
```latex
\begin{table}[ht]
    \centering
    \caption{Voltage and Temperature Squared Correlation}
    \label{tab:2}
    \begin{tabular}{*{2}c}
        \toprule
        Voltage, V, ($\pm 1$ V)  & Temperature, T \\ 
            \midrule
              1 & $(5 \pm 2) \times 10^{2} $  \\
              3 & $(30 \pm 3) \times 10^{2} $  \\
              5 & $(45 \pm 9) \times 10^{2} $  \\
              7 & $(7 \pm 2) \times 10^{3} $  \\
              8 & $(77 \pm 9) \times 10^{2} $  \\
              10& $(9 \pm 1) \times 10^{3} $ \\
        \bottomrule
    \end{tabular}
\end{table}
```
which results in

![](https://github.com/CianLM/labtex/raw/master/figures/readmetable2.png)

For a figure, a similar process occurs. Given any figure object, such as the one generated by `linreg.plot()` above, it can be added with the `doc.add_figure` method. For example, one could do
```python
squarereg = LinearRegression(voltages,temperatures**2)
fig = squarereg.plot(xlabel="Voltage, V", ylabel="Temperature Squared, $T^2$", title="Voltage and Temperature Squared Plot")
doc.add_figure(fig, caption="Linear Regression of Voltage and Temperature Squared")
# labtex: Wrote to 'figures/graph1.png'.
```
This generates the figure below and saves it to `figures/graph1.png` where the `1` will increment with each figure you add.

![](https://github.com/CianLM/labtex/raw/master/figures/graph3.png)

If you want a different filename, you can specify with the `filename=...` argument and if you want your graphs elsewhere than `figures/`, you may change `Document.graphfolder` at your convenience.

Once you have added all your tables and graphs to the `doc` object, you may save this file as shown below. The default write directory is `tex/` relative to root. This directory is also customizable with `Document.texfolder`.

```python
doc.save()
# labtex: Wrote to 'tex/report.tex'.
```

You can then edit the generated `.tex` file freely, and run the python file again at any point to add/edit/remove tables and figures. labtex will automatically update the `.tex` file with your changes.

By default, tables will be inserted after a line starting with `%labtex-tables` and figures will be inserted after a line starting with `%labtex-figures`. If you want to change the comments labtex looks for, you can edit the `Document.tablemarker` and `Document.figuremarker` attributes respectively. If such comments are not found, tables/figures will be inserted at the very end of the document. 

For more advanced usage and guides (such as the creation of the plot below) see the [Documentation](https://www.cianlm.dev/labtex).

![](https://github.com/CianLM/labtex/raw/master/figures/coherence.png)
## Contributions

This package is under active development. Feel free to submit a pull request or reach out with feature suggestions.


CHANGELOG.md
## Changelog

# Looking Forward

- CSV data import workflow
- Numpy input integration
- Interface with statistical packages for non-linear regression

# v0.6.1
- Fixed inverse tangent error propagation
- Added conda publishing workflow
- Added predict method to Regression objects
# v0.6.0
- Reworked the document object to be more versatile and to allow **live editing** of the document
- Improved generation of unit axis labelling
- Added numpy integration for instantiating MeasurementLists
- Added nonlinear regression
- Added parameter estimation region visualisation to plotting
- Improved unittesting
# v0.5.0
- New extensive ![Documentation](https://www.cianlm.dev/labtex)! Examples and guides to be added in future updates.
- Improved plot appearance with latex-style font
- Fixed table bugs when dimensionless quantities are involved
# v0.4.3
- Fixed negative error propagation bug
# v0.4.1
- Documentation/guides coming very soon
- Fixed exponential and inverse tangent error propagation
- Fixed other minor bugs
# v0.4.0

New Features:
- Unit conversion of measurements between arbitrary units e.g. J to eV, mm^2 to m^2.
- Massive overhaul of unit parsing and `MeasurementList`s for extensibility and consistency.
- Added "/" support e.g. "m/s" parses the same as "m s^-1"
- Added element-wise uncertainties in `MeasurementList`s
- Added eV and unit conversions eg. eV to Joule
- Added support for more prefixes up to 10^{± 12} (as Peta breaks parsing of Pascals)

Other changes include:
- Added shorthand classes `M` and `ML` for `Measurement` and `MeasurementList` classes
- Exposed the `plt` plot object to the end user for customization
- Simplified and standardized printing of uncertainties
- Fixed Linear Regression statistical uncertainties
- Modularized testing
- Bug fixes of all kinds



labtex/document.py
import subprocess
import regex as re
from labtex.linear import LinearRegression
from labtex.measurementlist import MeasurementList

from typing import Any, List, Union

import hashlib
import os

from labtex.nonlinear import NonlinearRegression
from matplotlib.pyplot import Figure

class Document:
    "A class for LaTeX template document creation with tables and graphs already inserted."
        # Customise these folders and the templates as you wish.
    texfolder = "tex/"
    figurefolder = "figures/"
    tablemarker = "%labtex-tables"
    figuremarker = "%labtex-figures"

    template = r"""\documentclass[]{article}

\title{!title}
\author{!author}

\usepackage{amsmath,amssymb,amsfonts,amsthm,physics,graphicx,geometry,enumitem,booktabs}

\begin{document}

\maketitle

\begin{abstract}

\end{abstract}

\section{Introduction}

\section{Theory}

\section{Method}

\section{Results}
!table
!tablemarker

\section{Data Analysis and Discussion}
!graph
!figuremarker

\section{Conclusion}

\section*{References}

\end{document}
"""


    tabletemplates = {
        "default": 
r"""\begin{table}[ht]!regex
    \centering
    \caption{!caption}
    \label{tab:!label}
    \begin{tabular}{!columns}
        \toprule
        !data
        \bottomrule
    \end{tabular}
\end{table}!table"""
    }

    graphtemplates = {
        "default": r"""
\begin{figure}[ht]!regex
    \centering
    \includegraphics[width=!width\textwidth]{!filename}
    \caption{!caption}
    \label{fig:!label}
\end{figure}!graph"""
    }
    
    def __init__(self, title : str, author : str, filename : str = "labdocument", silent : bool = False):
        "Initialise a LaTeX document with an title and an author."
        if os.path.exists(Document.texfolder + filename):
            self.document = open(Document.texfolder + filename).read()
        else:
            self.document = Document.template.replace("!title",title).replace("!author",author).replace("!tablemarker",Document.tablemarker).replace("!figuremarker",Document.figuremarker)
                # print("labtex: Overwriting existing file.")
        self.filename = filename
        self.tablenumber = 0
        self.graphnumber = 0
        self.silent = silent
        self.hashes = []
        self.unchanged_environments = {'tables': 0, 'figures': 0}

    def __repr__(self):
        return self.document

    def add_table(self,nameandsymbols : List[str], data : List[MeasurementList], \
        headers :List[str] = [], caption : str = "", label : str = "", style : str = "sideways"):
        """
        Add a table to the LaTeX document. 
        """
        hash_str = hashlib.sha1(str(data).encode("utf-8")).hexdigest()[:5]
        # id_str = f'%labtex-table-{self.tablenumber + 1}-{hash_str}'
        id_str = f'%labtex-table-{self.tablenumber + 1}'
        regex = re.compile('begin\{table\}\[[\w]{0,2}\]' + id_str + '.*?end{table}', re.DOTALL)
        
        self.hashes += [id_str]
        match = re.search(regex,self.document)
        table_code = self.table_code(nameandsymbols,data,headers,caption,label,style).strip()
        if match: # Replace table with the new table
            if (('\\' + match.group(0)) == table_code.replace('!table','')):
                # not self.silent and print("labtex: Figure code unchanged.")
                self.unchanged_environments['tables'] += 1
                return
            not self.silent and print("labtex: Updating table " + id_str)
            self.document = self.document.replace('\\' + match.group(0), table_code)
        else:
            key = Document.tablemarker
            not self.silent and print("labtex: Adding table " + id_str)
            if self.document.find(key) != -1:
                self.document = self.document.replace(key,'\n' + table_code + '\n' + key)
            else:
                self.document = self.document.replace('\\end{document}', '\n' + table_code + '\n' + '\\end{document}')

    # Called by add_table to generate the LaTeX code for the table
    def table_code(self,nameandsymbols : List[str], data : List[MeasurementList], \
        headers :List[str] = [], caption : str = "", label : str = "", style : str = "sideways"):

        assert len(nameandsymbols) == len(data)
        assert all(len(data[0]) == len(line) for line in data)
        columns = len(data[0])

        table = Document.tabletemplates["default"]
        self.tablenumber += 1
        # not self.silent and print(f"labtex: Adding table {self.tablenumber} to Document '{Document.texfolder + self.filename}'.")

        # generate 5 digit hash of the data
        # table = table.replace("!regex",f'%labtex-table-{self.tablenumber}-{hashlib.sha1(str(data).encode("utf-8")).hexdigest()[:5]}')
        table = table.replace("!regex",f'%labtex-table-{self.tablenumber}')
        table = table.replace("!label", str(self.tablenumber) if label == "" else label)
        table = table.replace("!caption",caption)

        if not (all(isinstance(line, MeasurementList) for line in data)):
            raise Exception("Data Error: Data should be a list of MeasurementLists.")

        if(style == "sideways"):
            # table = table.replace("!columns", "*{" + str(1+columns) + "}c" )
            table = table.replace("!columns", f"c|{ 'c' * columns}" )
            if(headers != [] and len(headers) >= 2):
                table = table.replace("!data",
                fr"""{headers[0]} & \multicolumn{{{columns}}}{{c}}{{{headers[1]}}} \\
            \midrule !data"""
                )
            for i in range(len(data)):
                table = table.replace("!data",
                fr"""
            {nameandsymbols[i]}{ data[i].tableprint() } \\ !data"""
                )

        elif(style == "upright"):
            table = table.replace("!columns", "*{" + str(len(data)) + "}c" )
            for i in range(len(data)):
                table = table.replace("!data",
                fr"""{nameandsymbols[i]}{data[i].tableprint(novalues=True)} & !data"""
                )
            table = table.replace("& !data",
            r"""\\ 
            \midrule
            !data"""
            )
            tableprint = [ml.tableprint(nounits=True)[1:].split("&") for ml in data]
            indexfirst = [ [index[j] for index in tableprint] for j in range(len(tableprint[0]))]

            for index in indexfirst:
                table = table.replace("!data",
            rf""" {"&".join([*index])} \\
            !data""")

        else:
            raise Exception("Style Error: Only 'sideways' and 'upright' styles are supported.")


        table = table.replace("!data","")
        table = re.sub(r'\n[ \t]+\n', '\n', table)

        # self.document = self.document.replace("!table",table)
        return table

    def add_figure(self, fig : Figure, caption : str = "", label : str = "", width : float = 0.8, filename = ""):
        "Add figure to the LaTeX document."
        
        # Save figure to file
        if fig is not None:
            if (not os.path.exists(Document.figurefolder)):
                os.makedirs(Document.figurefolder)
            filename = f"graph{self.graphnumber + 1}.png" if not filename else filename
            fig.savefig(Document.figurefolder + filename)
            not self.silent and print(f"labtex: Wrote to '{Document.figurefolder + filename}'.")

        id_str = f'%labtex-figure-{self.graphnumber + 1}'
        regex = re.compile('begin\{figure\}\[[\w]{0,2}\]' + id_str + '.*?end{figure}', re.DOTALL)

        self.hashes += [id_str]
        match = re.search(regex,self.document)
        figure_code = self.figure_code(caption,label,width,filename).strip()
        if match: # Replace figure
            if (('\\' + match.group(0)) == figure_code.replace('!graph','')):
                # not self.silent and print("labtex: Figure code unchanged.")
                self.unchanged_environments['figures'] += 1
                return

            not self.silent and print("labtex: Updating figure " + id_str)
            self.document = self.document.replace('\\' + match.group(0), figure_code)
        else:
            key = Document.figuremarker
            print("labtex: Adding figure " + id_str)
            if self.document.find(key) != -1:
                self.document = self.document.replace(key,'\n' + figure_code + '\n' + key)
            else:
                self.document = self.document.replace('\\end{document}', '\n' + figure_code + '\n' + '\\end{document}')

        
    # Called by add_figure to generate the latex code for the figure
    def figure_code(self, caption : str = "", label : str = "", width : float = 0.8, filename = ""):
        graph = Document.graphtemplates['default']
        self.graphnumber += 1
        # graph = graph.replace("!regex",f'%labtex-graph-{self.graphnumber}-{hashlib.sha1(str(data).encode("utf-8")).hexdigest()[:5]}')
        graph = graph.replace("!regex",f'%labtex-figure-{self.graphnumber}')
        graph = graph.replace("!label",str(self.graphnumber) if label == "" else label)
        graph = graph.replace("!caption",caption)
        graph = graph.replace("!width",str(width))

        filename = f"graph{self.graphnumber}" if not filename else filename
        if(Document.figurefolder != '.'): # assuming the graph folder is a subfolder
            graph = graph.replace("!filename","../" + Document.figurefolder +  filename)
        else:
            graph = graph.replace("!filename", Document.figurefolder + filename)

        return graph

    def save(self, overwrite: bool = False):
        "Save the document to 'filename'."
        # Compile the tex file to check for errors
        if os.path.exists(Document.texfolder + self.filename):
            try:
                not self.silent and print(f"labtex: Compiling '{Document.texfolder + self.filename}' to check for errors.")
                subprocess.run(
                    ["pdflatex", "-halt-on-error", self.filename],
                    stderr=subprocess.DEVNULL,
                    stdout=subprocess.DEVNULL,
                    cwd=Document.texfolder,
                    check=True,
                    # suppress output

                )
            except:
                raise Exception("labtex: File failed to compile prior to saving. Save cancelled.")
        else:
            not self.silent and print(f"labtex: File '{Document.texfolder + self.filename}' does not exist. Creating new file.")
        # Remove tables and graphs that are not in the document
        # id_str = '(%labtex-table-\w{1,2}-\w{5})'
        id_str = '(%labtex-(?:table|figure)-\w{1,2})'
        regex = re.compile('(begin\{(table|figure)\}\[[\w]{0,2}\]' + id_str + '.*?end{(?:table|figure)})', re.DOTALL)
        match_all = regex.findall(self.document)
        # print(self.hashes)
        # print(len(match_all))
        for match, tof, id_str in match_all: # tof stands for 'table or figure'
            if id_str not in self.hashes:
                self.document = self.document.replace('\\' + match + '\n', '')
                not self.silent and print('labtex: Removing', id_str)

        # Display number of unchanged environments
        if not self.silent:
            self.unchanged_environments['tables'] != 0 and print(f"labtex: {self.unchanged_environments['tables']} unchanged tables.")
            self.unchanged_environments['figures'] != 0 and print(f"labtex: {self.unchanged_environments['figures']} unchanged figures.")

        self.document = self.document.replace("!table","").replace("!graph","")

        if(not os.path.exists(Document.texfolder)):
            not self.silent and print("labtex: Creating folder '" + Document.texfolder + "'.")
            os.makedirs(Document.texfolder)
        if(os.path.exists(Document.texfolder + self.filename)):
            not self.silent and print(f"labtex: '{Document.texfolder + self.filename}' already exists. { 'Overwriting.' if overwrite else 'Use `save(...,overwrite=True)` to overwrite.'}")
            if(not overwrite):
                return

        with open(Document.texfolder + self.filename,'w') as outputfile:
            outputfile.write(self.document)

        not self.silent and print(f"labtex: Wrote to '{Document.texfolder + self.filename}'.")

# Update document procedure:
# 1. For each table/graph call:
#   1.1. Check if table/graph is in file
#   1.2. If not, add it to bottom
#   1.3. If so, update/replace it
#   1.4. Remove tables/graphs that are not in the document
# 2. Save file


# Intended UX:
# 1. Create document
# 2. Add tables/graphs
# 3. Save document
# 4. Change contents of tables/graphs
# 5. This should keep their positions in the document even if the content changes
